### DISCUSSION TITLE
add_mesh becomes very slow all of a sudden after many calls to Plotter

### DISCUSSION THREAD

**Discussion by leo2r** [pyvista/pyvista#3331] [Q&A]
Hi all,

I'm wondering if you've seen this behaviour before. I'm using PyVista to create the environment to train a reinforcement learning agent. I have a function to create the image that gets fed to the agent as an observation, which uses a PyVista ```Plotter``` object. My function works well and quickly for the first several dozens of thousands of calls, but for some reason at some point becomes super slow (goes from taking about 0.02 seconds to 4-6 seconds). Below is the code with debugging statements to see which line exactly was causing the problem:

```python
pl = pv.Plotter(off_screen=True) # Plotter instance as a class attribute

def get_image(self, cpos=[0, 0, 0], croll=0):
        start = time.time()
        self.pl.clear()
        end = time.time()
        print(f'clear : {end - start}')

        start = time.time()
        self.pl.add_mesh(self.bone_mesh, color="lightblue", opacity=0.3) #style='wireframe',
        self.pl.add_mesh(self.screw1_mesh, color=True, opacity=1)
        self.pl.add_mesh(self.screw2_mesh, color=True, opacity=1)
        end = time.time()
        print(f'add_mesh : {end - start}')

        start = time.time()
        self.pl.camera.position = cpos
        self.pl.camera.focal_point = (0,0,0)
        self.pl.camera.roll = croll
        end = time.time()
        print(f'camera : {end - start}')

        start = time.time()
        img = self.pl.screenshot(None, return_img=True, window_size=(self.WINDOW_SIZE))
        end = time.time()
        print(f'screenshot : {end - start}')

        return img[..., :1] # Correct format from (WIDTH, HEIGHT) to (WIDTH, HEIGHT, 1)
```

And attached you can see the results (before and after). It shows that the ```add_mesh``` function goes from taking 0.008-9 seconds to 2-3 seconds after many calls.

Any idea what's going on here and how I can remedy this? My training FPS for the environment goes from 72FPS to 0-1FPS after a few minutes because of this :'(

Between calls to ```get_image```, I move the meshes around with the proposed actions from the agent as so:
```python
# Take action
self.screw1_mesh.translate(screw1_move_xyz, inplace=True)
self.screw1_mesh.rotate_x(screw1_rot_xz[0], inplace=True)
self.screw1_mesh.rotate_z(screw1_rot_xz[1], inplace=True)
      
self.screw2_mesh.translate(screw2_move_xyz, inplace=True)
self.screw2_mesh.rotate_x(screw2_rot_xz[0], inplace=True)
self.screw2_mesh.rotate_z(screw2_rot_xz[1], inplace=True)
```

Quick at the start:
<img width="246" alt="Screenshot 2022-09-15 at 14 07 55" src="https://user-images.githubusercontent.com/27723482/190399762-605117e2-98c7-4013-b163-316c6d97f5be.png">

Slow after several dozens of thousands of calls:
<img width="250" alt="Screenshot 2022-09-15 at 14 07 00" src="https://user-images.githubusercontent.com/27723482/190399772-c8aa6609-fa2c-4a1b-bdeb-962b7142ab23.png">

**Comment by leo2r** [comment]
Another point I should make is that the exact same code works fine on my MacBook Pro, but this problem occurs on a Ubuntu virtual machine...

**Comment by leo2r** [comment]
@banesullivan @akaszynski @adeak Hi guys, can any of you have a look at this or push it to someone who can? 🙏🏽 Thanks :)

**Reply by banesullivan** [reply]
Sorry for our silence here... this is an interesting problem. I'm actually wondering if you could try this out on `main` to see if the issue persists after @akaszynski's refactor in #2964 and his addition of composite mappers that helped speed up `add_mesh` in #2941.

After which, if the issue persists, could you work with us to make a reproducible example of this behavior?

**Reply by leo2r** [reply]
My turn to apologise for the delayed response. I am on main already 😓 So unfortunately I think the issue is still here despite @akaszynski's refactoring :( I will try now to create a reproducible example! Will get back to you!

**Reply by leo2r** [reply]
So I seem to have found a reproducible example of the behaviour! See below..

```python
import pyvista as pv
import time 

pl = pv.Plotter(off_screen=True)
sphere = pv.Sphere()
cube = pv.Cube()
another_cube = cube.copy(deep=True)
flag = True

while flag:
    start = time.time()
    pl.clear()
    pl.add_mesh(sphere, color="lightblue", opacity=0.3)
    pl.add_mesh(cube, color=True, opacity=1)
    pl.add_mesh(another_cube, color=True, opacity=1)
    finish = time.time()
    print(finish - start)
    img = pl.screenshot(None, return_img=True)
```

It runs fine for around 4-6 minutes, taking about 0.006 seconds but all of a sudden jumps up to around 3 seconds. Seems to have to do with the ```screenshot()``` function as omitting that line allowed the code to run fine for well over 10 minutes..

<img width="159" alt="Screenshot 2022-09-26 at 16 58 00" src="https://user-images.githubusercontent.com/27723482/192311794-6d1912ca-9cb3-41a0-bb3b-fdf1168eadda.png">

I'm on Ubuntu 20.04.5 LTS, using PyVista's up-to-date main branch from here. Let me know if you need any more info!

**Reply by banesullivan** [reply]
Thanks for creating a reproducible example, @leo2r! We'll try to look into what is going wong here

**Reply by adeak** [reply]
I've been running your script for a bit more than 4 hours straight on a debian system, head very slightly ahead of main. I still see a steady 0.03 seconds per iteration, just like at the start. Even if I consider that my system is overall 5 times slower than yours, I should be seeing something by now even if we assume that you need a given number of iterations for problems to show up.

**Reply by leo2r** [reply]
@adeak Damn that's odd... sorry about that! I'll look into it again tomorrow :(

**Reply by adeak** [reply]
Well, if it's reproducible on your system then it's still something, so there's nothing to be sorry about. It just makes figuring this out harder.

**Reply by leo2r** [reply]
Ok I just double checked and it again took less than 10 minutes to reach the point where performance massively slows down. I'm not sure what to do? :(

**Reply by akaszynski** [reply]
Ran this locally for 18 hours with no issues or slowdown. Ran at 0.03219103813171387 per call.

```py
from tqdm import tqdm
import pyvista as pv
import time 

pl = pv.Plotter(off_screen=True)
sphere = pv.Sphere()
cube = pv.Cube()
another_cube = cube.copy(deep=True)
times = []

try:
    for _ in tqdm(range(10000000)):
        start = time.time()
        pl.clear()
        pl.add_mesh(sphere, color="lightblue", opacity=0.3)
        pl.add_mesh(cube, color=True, opacity=1)
        pl.add_mesh(another_cube, color=True, opacity=1)
        telap = time.time() - start
        img = pl.screenshot(None, return_img=True)
        times.append(telap)
except KeyboardInterrupt:
    breakpoint()
```

```
>>> times
... 0.03210020065307617, 0.03219103813171387, 0.03235578536987305, 0.03 ...
>>> len(times)
1344431
>>> np.max(times)
0.9
>>> np.mean(times)
0.03218189450235425
>>> times[-1]
0.03295397758483887
```

```py
>>> pv.Report()
```

```
--------------------------------------------------------------------------------
  Date: Tue Sep 27 10:59:33 2022 MDT

                OS : Linux
            CPU(s) : 6
           Machine : x86_64
      Architecture : 64bit
               RAM : 62.6 GiB
       Environment : Python
       File system : ext4
        GPU Vendor : NVIDIA Corporation
      GPU Renderer : Quadro P2000/PCIe/SSE2
       GPU Version : 4.5.0 NVIDIA 470.141.03

  Python 3.8.10 (default, Jun 22 2022, 20:18:18)  [GCC 9.4.0]

           pyvista : 0.37.dev0
               vtk : 9.1.0
             numpy : 1.23.3
           imageio : 2.22.0
           appdirs : 1.4.4
            scooby : 0.5.12
        matplotlib : 3.6.0
           IPython : 8.5.0
          colorcet : 3.0.0
           cmocean : 2.0
        ipyvtklink : 0.2.2
             scipy : 1.9.1
        itkwidgets : 0.32.3
              tqdm : 4.64.1
            meshio : 5.3.4
        jupyterlab : 3.4.7
         pythreejs : Version unknown
--------------------------------------------------------------------------------
```

**Reply by leo2r** [reply]
Damn... this is frustrating. The only difference I see between @akaszynski and my report is that I'm a few minor releases behind on Numpy, imageio etc. and am using Python 3.9.7. Any suggestions on how to proceed with this issue? Thank you!

**Reply by adeak** [reply]
@leo2r the only thing I can think of is for you to run the debugging snippet of @akaszynski, and when it slows down after a few minutes press ctrl+c and try to investigate the environment. Is there large memory use in the OS? Are other programs slow? Is the same python session slow if you try to run things? If you start another interpreter running the same script, does it start fast and slow down again or does it start slow? I have no clear plan of what to do in this situation, but currently you're the only one who can reproduce the problem so only you can debug this.
